pkg/provider/aws/resources/iam.go

```diff
@@ -147,6 +147,30 @@ type (
 	}
 )
 
+type RoleCreateParams struct {
+	AppName string
+	Name    string
+	Refs    []core.AnnotationKey
+}
+
+// Create takes in an all necessary parameters to generate the Roles name and ensure that the Role is correlated to the constructs which required its creation.
+func (role *IamRole) Create(dag *core.ResourceGraph, params RoleCreateParams) error {
+	role.Name = roleSanitizer.Apply(fmt.Sprintf("%s-%s", params.AppName, params.Name))
+	role.ConstructsRef = params.Refs
+
+	existingRole := dag.GetResource(role.Id())
+	if existingRole != nil {
+		return fmt.Errorf("iam role with name %s already exists", role.Name)
+	}
+
+	dag.AddResource(role)
```

Comment:
This pattern happens a lot, maybe add a method on the dag to reduce copy/paste?

Reply:
yeah going to add this to the same ticket